## Introduction

It can be challenging to distinguish signal from noise in biomedical datasets, and machine learning methods are particularly hampered with the amount of available training data is small.
Incorporating biomedical knowledge into machine learning models can reveal patterns in noisy data [@doi:10.1038/nrg.2017.38] and aid model interpretation [@doi:10.1016/j.cell.2018.05.056].
Biological knowledge can take many forms, including genomic sequences, pathway databases, gene interaction networks, and knowledge hierarchies such as the Gene Ontology [@doi:10.1093/nar/gky1055].
However, there is often no canonical way to encode these structures as real-valued predictors.
Modelers must creatively decide how to encode biological knowledge that they expect will be relevant to the task.

Biomedical datasets often contain more input predictors than data samples [@doi:10.1109/JPROC.2015.2494198; @arxiv:1611.09340].
A genetic study may genotype millions of single nucleotide polymorphisms (SNPs) in thousands of individuals, or a gene expression study may profile the expression of thousands of genes in tens of samples.
Thus, it can be useful to include prior information describing relationships between predictors to inform the representation learned by the model.
This contrasts with non-biological applications of machine learning, where one might fit a model on millions of images [@doi:10.1109/CVPR.2009.5206848] or tens of thousands of documents [@url:https://www.aclweb.org/anthology/P11-1015/], making inclusion of prior information unnecessary.

We review approaches that incorporate external information about the structure of desirable solutions to learn from biomedical data.
One class of commonly used approaches learns a representation that considers the context of each base pair from raw sequence data.
For models that operate on gene expression data or genetic variants, it can be useful to incorporate networks or pathways describing relationships between genes.
We also consider other examples, such as neural network architectures that are constrained based on biological knowledge.

In addition to the methods surveyed here, there are many complementary options for utilizing heterogeneous sources of biomedical data.
These include feature extraction or representation learning prior to modeling, and/or other data integration methods that do not necessarily involve customizing the model itself.
Many of these methods have been covered extensively elsewhere [@doi:10.3389/fgene.2019.00381; @doi:10.1016/j.inffus.2018.09.012].

