## Sequence-Constrained Models

1. Use in regulatory biology (methylation, histone marks, TFs, splicing)
    a. Early examples
        * 10.1126/science.1254806 (can cite as an early example of using
          hand-engineered features as input to a 2-layer MLP, preceded most
          CNN examples on sequence data -> splicing exon inclusion prediction)
        * 10.1038/nbt.3300 (CNN for prediction of DNA/RNA binding protein
          sequence specificity)
        * 10.1038/nmeth.3547 (CNN for multi-task variant effect prediction
          from noncoding genome, one of the earlier examples of CNN on genomic
          sequence data)
        * 10.1101/gr.200535.11 (CNN for prediction of chromatin accessibility
          from DNA sequence, evaluated to predict effects of SNPs)
    b. More recent examples
        * 10.1101/737981 (CNN for prediction of TF binding probability and
          binding strength from sequence, used to identify and interpret
          cis-regulatory code i.e. order and spacing of TF binding motifs,
          no need for strong modeling assumptions + allow lots of context
          ~1000 base windows)
        * 10.1016/j.cell.2019.04.046 (prediction of alternative polyadenylation
          from sequence, used to predict probability of poly-A cleavage and
          engineer new PASs)
        * DNA methylation models typically operate on microarray beta values
          (proportion of DNA methylated at each CpG locus), so these wouldn't
          be considered sequence-constrained. An exception: DeepSignal
          (10.1093/bioinformatics/btz276) - CNN that works on electrical signals
          around CpG sites from Nanopore sequencing reads, to predict 5mC or 6mA
          methylation status
        * 10.1101/329334 and 10.1093/bioinformatics/bty612: prediction of gene
          expression (or differential gene expression) from histone modification
          signals. Signals are discretized using bins (no raw sequence input)
          but bidirectional LSTM considers context in both directions and allows
          for learning complex histone mark dependencies.
2. Use in variant calling
    * 10.1038/nbt.4235 (actually works on images of read pileups, but uses sequence context)
    * 10.1038/s41467-019-09025-z (single molecule sequencing: PacBio/ONT, works on featurization of sequence
      defined in paper, outperforms ^ on these)
    * 10.1101/097469 (similar to DeepVariant, but works on raw sequences rather than images,
      \+ less preprocessing/claims to learn more info directly from sequences)
    * 10.1101/601450 (calling of short indels using sequence information)
3. Use in mutation detection
    * 10.1038/s41467-019-09027-x (mutation detection in tumors)
4. Use in CRISPR guide selection
    * 10.1371/journal.pcbi.1005807 (shows that a wide variety of sequence attributes
      are useful for predicting cleavage efficiency, didn't use contextual info?)
    * 10.1038/nbt.3437, 10.1038/s41551-017-0178-6 (models for predicting on-target efficacy
      and off-target effects, uses contextual info about sequence i.e. order 2+ features)
    * 10.1038/nbt.4061 (CNN-based model for Cpf1 activity prediction,
      convolutions of length 5 on one-hot encoded sequence)
    * 10.1186/s13059-018-1459-4 (similar to ^ but for Cas9, also uses ChIP-seq
      data)
    * 10.1101/636472 (similar to ^)
    * 10.1101/505602 (uses STRING -> gene network-based features, in addition
      to sequence features, for on-target efficacy prediction)
