## Sequence-Constrained Models



(a bit about how a model might be constrained by its sequence goes here)
(maybe something about how sequences are one-hot encoded or how convolutions might apply to them? depends on who we expect the audience of this paper to be)

### Applications in regulatory biology

Many of the first applications of deep learning to biological sequence data were in regulatory biology.
Early neural network models primarily used hand-engineered sequence features as input to a fully connected neural network [@doi:10.1126/science.1254806; another example here?].
As convolutional neural network (CNN) approaches matured for image and text processing, researchers were able to adapt similar ideas to biological sequence data.
Example early applications of CNNs on sequence data include prediction of binding protein sequence specificity from DNA or RNA sequence [@doi:10.1038/nbt.3300], prediction of variant effects from noncoding DNA sequence [@doi:10.1038/nmeth.3547], and prediction of chromatin accessibility from DNA sequence [@doi:10.1101/gr.200535.115].

Recent sequence models have taken advantage of hardware advances and methodological innovation to incorporate more sequence context and rely on fewer modeling assumptions.
BPNet, a CNN used to predict transcription factor binding profiles from raw DNA sequences, was able to accurately map known locations of binding motifs in mouse embryonic stem cells [@doi:10.1101/737981].
The BPNet model considers 1000 base pairs of context around each position when predicting binding probabilities, which is important because (blah blah).
APARENT [@doi:10.1016/j.cell.2019.04.046] is a CNN used to predict alternative polyadenylation (APA) from a training set of over 3 million synthetic APA reporter sequences.

Models that consider sequence context have also been applied to make predictions from epigenetic data.
To make predictions (?? or as labels? check on this) from DNA methylation data, microarray beta values are typically used as input, which quantify the proportion of DNA methylated at each locus profiled by the study.
An exception is DeepSignal [@doi:10.1093/bioinformatics/btz276], which is a CNN that uses electrical signals from Nanopore sequencing to predict 5mC or 6mA DNA methylation status.
Deep learning models have also been used to predict gene expression from histone modifications [@doi:10.1101/329334; @doi:10.1093/bioinformatics/bty612].
In this case, a neural network model consisting of long short-term memory (LSTM) units was used to encode the long-distance interactions of histone marks in both the 3' and 5' genomic directions.

### Applications in variant calling and mutation detection

* 10.1038/nbt.4235 (actually works on images of read pileups, but uses sequence context)
* 10.1038/s41467-019-09025-z (single molecule sequencing: PacBio/ONT, works on featurization of sequence
  defined in paper, outperforms ^ on these)
* 10.1101/097469 (similar to DeepVariant, but works on raw sequences rather than images,
  \+ less preprocessing/claims to learn more info directly from sequences)
* 10.1101/601450 (calling of short indels using sequence information)
* 10.1038/s41467-019-09027-x (mutation detection in tumors)

### Applications in CRISPR guide selection

* 10.1371/journal.pcbi.1005807 (shows that a wide variety of sequence attributes
  are useful for predicting cleavage efficiency, didn't use contextual info?)
* 10.1038/nbt.3437, 10.1038/s41551-017-0178-6 (models for predicting on-target efficacy
  and off-target effects, uses contextual info about sequence i.e. order 2+ features)
* 10.1038/nbt.4061 (CNN-based model for Cpf1 activity prediction,
  convolutions of length 5 on one-hot encoded sequence)
* 10.1186/s13059-018-1459-4 (similar to ^ but for Cas9, also uses ChIP-seq
  data)
* 10.1101/636472 (similar to ^)
* 10.1101/505602 (uses STRING -> gene network-based features, in addition
  to sequence features, for on-target efficacy prediction)
