## Network- and Pathway-Constrained Models

### Applications in bulk transcriptomics

* 10.1016/j.cels.2019.04.003 (pathways -> transfer learning for rare diseases)
* 10.1038/s41540-019-0086-3 (PPI + pathways -> phenotype prediction/cancer subtyping)
* https://arxiv.org/pdf/1906.10670 (HumanBase -> AML drug response prediction)
* 10.1093/bioinformatics/bty429 (HINT PPI)
* 10.1038/s41598-017-03141-w (PPI -> mutated cancer gene detection)
* 10.1371/journal.pcbi.1006657 (PPI + network optimization -> cancer outcome prediction)
* 10.1186/s12859-018-2500-z (pathways -> NN structure, for predicting patient outcomes in GBM)
* 10.1186/s12859-017-1984-2 (gene regulatory network -> NN structure, for patient outcomes in
 kidney transplantation and ulcerative colitis)
* 10.1038/s41598-018-19635-0 (gene regulatory network -> group lasso, same datasets as ^)
* 10.1093/bioinformatics/bty945 (gene regulatory network -> NN structure, for predicting
 gene expression given gene knockouts)

### Applications in single cell transcriptomics

* 10.1038/s41592-019-0456-1 (pathways -> cell type deconvolution/pathway-level eQTLs)
* 10.1093/nar/gkx681 (PPI; protein-DNA interactions)
* 10.1101/544346 (PPI)

### Applications in genetics

1. Genotype-phenotype associations (GWAS/eQTL)
   * 10.1093/bioinformatics/btu293
   * 10.1093/bioinformatics/btx677 (multivariate version of ^)
   * 10.1111/biom.13072 (phenotype prediction using variant and GE data)
2. Using pathways as a prior to study SNP-SNP or gene-gene interactions
 (not sure if this is truly a constraint on the model, more just a
  strategy to reduce the number of hypothesis tests)
   * 10.1101/182741, 10.1371/journal.pgen.1006973 (pathways + GWAS data -> GIs)
   * 10.1371/journal.pgen.1006516 (pathways + GWAS data -> GIs)
3. Using SNP-SNP or gene-gene networks to reduce hypothesis testing
  burden for detecting GIs
   * can find some examples if we decide this is a direction worth going
