## Other constrained models

1. Ontology-constrained models: use knowledge hierarchies to inform the model
    a. Phenotype prediction
        * 10.1038/nmeth.4627 (gene ontology -> NN structure, to predict effects of mutations on growth in yeast)
    b. Function prediction
        * 10.1093/bioinformatics/btx624 (GO hierarchy relationships -> NN structure, for function prediction)
        * 10.1093/bioinformatics/btx252 (PPI network + tissue ontology -> function prediction)

2. Otherwise constrained models: use other biological knowledge to inform the model (??)
    a. Cell cycle information
        * 10.1038/nbt.3102
        * 10.1101/526848
       (in principle the denoising method could be generalized to other gene
        sets, but here they used cell cycle-relevant gene sets and emphasized
        the utility of this)
    b. Circadian rhythms: 10.1073/pnas.1619320114 (circular node autoencoder
      for modeling periodic gene expression)
    c. TAD/3D chromatin structure information? Can probably find some examples
      of this
