## Other constrained models

### Ontology-constrained models

1. Phenotype prediction
    * 10.1038/nmeth.4627 (gene ontology -> NN structure, to predict effects of mutations on growth in yeast)
2. Function prediction
    * 10.1093/bioinformatics/btx624 (GO hierarchy relationships -> NN structure, for function prediction)
    * 10.1093/bioinformatics/btx252 (PPI network + tissue ontology -> function prediction)

### Otherwise constrained models

1. Phylogeny-regularized models
    * 10.3389/fmicb.2018.01391: tree-regularized LMM for predicting an outcome based on microbiome sequencing data, application to predicting age based on gut microbiome
    * 10.3389/fmicb.2018.03112: generalizes ^ to sparse modeling/feature selection, applied to prediction of caffeine intake from gut microbiome and smoking status from upper respiratory tract microbiome
    * 10.1101/519892: tumor clonal deconvolution using phylogeny-regularized NMF model
2. Circadian rhythms
    * 10.1073/pnas.1619320114: circular node autoencoder for modeling periodic gene expression
3. Drug-gene interactions (this may belong in the network section):
    * https://arxiv.org/abs/1808.06603 (uses PPI network and drug-gene interactions from GDSC to bias the structure of a tree ensemble, for drug sensitivity prediction)
